discovery/spec.md

```diff
@@ -778,135 +782,145 @@ upon return of the `202 Accepted` from the original request.
-of the request, the processing of each Service in the request MUST be
-applied as if all previous Services in the request were already completed.
-This means that attribute uniqueness checking, such as on the `name`
-attribute, MUST be done on the new state of all Services as if all previously
-specified Services in the incoming request were already successfully processed.
-
-For example, if a Discovery Endpoint has a Service called `MyService`, it is
-possible to rename that Service to `YourService` and then to add, or update,
-another Service to use `MyService` as its `name`.
-
-Likewise, while it might be less than optimal, it is possible for
-a request to update a Service multiple times if the same `id` is used more
-than once within a request.
-
-The follow responses are defined by this specification:
-- `200 OK` if all the specified Services were processed successfully
-  - The HTTP Response Body MUST include an array of the Service values
-    resulting from the successful processing of the request. The order of
-    those values MUST match the order of the Services in the request. The
-    response MAY include attributes and values which were added by the
-    Discovery Endpoint.
+contain all Services in the request or fail. The body of the request message
+MUST contain an array of zero or more Service entries.
+
+The following rules apply to processing the Services specified in the request:
+- any error processing the request MUST result in the request having no
+  effect on the Discovery Endpoint.
+- a Service (as identified by its `id`) MUST NOT appear more than once in the
+  request.
+- any existing Service (as identified by its `id`) MUST be completely
+  replaced by the Service definition in the request, except as noted below.
+  This means that any missing property in the incoming request MUST be
+  treated as a request to delete the property from the existing Service
+  definition.
+- if an incoming Service does not contain an `id` then the Discovery Endpoint
+  MUST assign an appropriate value. The value MUST be globally unique.
+- if there is no existing Service with the given `id` then a new Service MUST
+  be created with that `id`.
+- if an incoming Service has an `epoch` value then the request MUST fail if
+  that value is not greater than the current Service's `epoch` value.
+- if an incoming Service does not have an `epoch` value then the Service's
+  `epoch` value MUST be set to a valid value of the Discovery Endpoint's
+  choosing. This means that if there is an existing Service with the same
+  `id` then the value MUST be set to a value greater than the current value.
+- aside from `id` and `epoch`, unless there is an out of band agreement,
+  all mandatory attributes MUST be present in the request, and a Discovery
+  Endpoint MUST reject requests that are missing such attributes.
+- the Discovery Endpoint MUST ignore attempts to modify read-only or immutable
```

Comment:
I'm not sure we need to say that you can't modify an `id`. How could you, given that the `id` is used to identify the service you're trying to update?

Reply:
To me it's not so much about repeating what "immutable" means, but rather what to do in the event that someone tries... ignore vs fail. And in this case I'm suggesting it ignore it. I think that's easier to think about for URL, but for ID... might be unnecessary for POST, but I think it applies for PUT.

Reply:
Ah... because the id is present in both the service object and in the URL? That's already required in the PUT request via

> The body of the request message MUST contain a Service with an id attribute matching the {id} in the path.

... so I don't think it needs repeating, and in the case of the POST, such an attempt can't even be expressed as far as I can tell.

Reply:
ok I removed it.